In [1]:
import pandas as pd
import sqlite3

# Connexion à la base de données
conn = sqlite3.connect('/Users/Radja/Downloads/batch_1.db')

# Récupération des données de la table aeroports
aéroports = pd.read_sql_query("SELECT * from aeroports", conn)

# Fermeture de la connexion à la base de données
conn.close()

# Affichage des données
aéroports.head()

CODE IATA                               NOM         LIEU PAYS  \
0       MCT      Muscat International Airport       Muscat   OM   
1       SOU               Southampton Airport  Southampton   GB   
2       PNH  Phnom Penh International Airport   Phnom Penh   KH   
3       BLR  Kempegowda International Airport    Bangalore   IN   
4       FFD                      RAF Fairford     Fairford   GB   

             LONGITUDE             LATITUDE  HAUTEUR  \
0   58.284400939941406   23.593299865722656     48.0   
1  -1.3567999601364136    50.95029830932617     44.0   
2   104.84400177001953   11.546600341796875     40.0   
3            77.706299              13.1979   3000.0   
4  -1.7900300025900002        51.6822013855    286.0   

   PRIX RETARD PREMIERE 10 MINUTES  \
0                             6695   
1                             1061   
2                             2402   
3                             8641   
4                             6883   

   PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES  
0                                              361  
1                                               71  
2                                              909  
3                                              550  
4                                              870

In [2]:
aéroports.describe()

HAUTEUR  PRIX RETARD PREMIERE 10 MINUTES  \
count   321.000000                       321.000000   
mean    711.283489                      5167.186916   
std    1232.281912                      2820.141750   
min       6.000000                       105.000000   
25%      59.000000                      2738.000000   
50%     220.000000                      5499.000000   
75%     685.000000                      7489.000000   
max    7630.000000                      9877.000000   

       PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES  
count                                       321.000000  
mean                                        502.947040  
std                                         281.834161  
min                                          11.000000  
25%                                         267.000000  
50%                                         510.000000  
75%                                         738.000000  
max                                         995.000000

Pour la variable "hauteur" (en pieds), la moyenne est de 711,28 pieds, avec une forte dispersion (écart-type élevé de 1232,28 pieds). On remarque également que la valeur minimale est de 6 pieds et que la valeur maximale est de 7630 pieds, ce qui suggère qu'il y a une grande variabilité des hauteurs des aéroports dans cette base de données.
Pour la variable "prix retard première 10 minutes", la moyenne est de 5167,19 dollars, avec un écart-type de 2820,14 dollars. On remarque également que la valeur minimale est de 105 dollars et que la valeur maximale est de 9877 dollars. Les quartiles indiquent que la moitié des observations sont comprises entre 2738 et 7489 dollars.
Pour la variable "prix retard pour chaque minute après 10 minutes", la moyenne est de 502,95 dollars, avec un écart-type de 281,83 dollars. On remarque que la dispersion est similaire à celle de la variable précédente. Les quartiles indiquent que la moitié des observations sont comprises entre 267 et 738 dollars.
Ces informations suggèrent qu'il y a une grande variabilité dans les hauteurs et les coûts de retard des aéroports de cette base de données. 

In [4]:
duplicate_iata_codes = aéroports['CODE IATA'].duplicated(keep=False)
duplicate_count = sum(duplicate_iata_codes)
print("Number of duplicate CODE IATA values:", duplicate_count)
aéroports.drop_duplicates(subset='CODE IATA', keep='first', inplace=True)

Number of duplicate CODE IATA values: 3


In [3]:
import pandas as pd
import sqlite3

# Connexion à la base de données
conn = sqlite3.connect('/Users/Radja/Downloads/batch_1.db')

# Récupération des données de la table aeroports
vols = pd.read_sql_query("SELECT * from vols", conn)

# Fermeture de la connexion à la base de données
conn.close()

# Affichage des données
vols.head()

IDENTIFIANT   VOL  CODE AVION AEROPORT DEPART AEROPORT ARRIVEE  \
0       651373  2317  28fa5df4d7             TIA              DSS   
1      4153042  1755  4b31cd6358             MPM              BOH   
2      3617513   804  f3e0769c08             TRN              TNA   
3      1829601   263  2ba258e5c4             BRU              SXF   
4      1250864  4511  11e7ffe99f             VCE              AAL   

   DEPART PROGRAMME  HEURE DE DEPART  RETART DE DEPART  \
0              1410           1411.0               1.0   
1              1455           1449.0              -6.0   
2              1540           1607.0              27.0   
3              2226           2219.0              -7.0   
4              1730           1719.0             -11.0   

   TEMPS DE DEPLACEMENT A TERRE AU DECOLLAGE  DECOLLAGE  ...  ANNULATION  \
0                                       20.0     1431.0  ...           0   
1                                       15.0     1504.0  ...           0   
2                                        9.0     1616.0  ...           0   
3                                       14.0     2233.0  ...           0   
4                                       14.0     1733.0  ...           0   

   RAISON D'ANNULATION  RETARD SYSTEM  RETARD SECURITE  RETARD COMPAGNIE  \
0                 None            NaN              NaN               NaN   
1                 None            NaN              NaN               NaN   
2                 None            NaN              NaN               NaN   
3                 None            NaN              NaN               NaN   
4                 None            NaN              NaN               NaN   

   RETARD AVION  RETARD METEO       DATE  NIVEAU DE SECURITE  \
0           NaN           NaN  10/4/2016                  10   
1           NaN           NaN  16/7/2018                  10   
2           NaN           NaN   6/1/2018                  10   
3           NaN           NaN   9/1/2016                  10   
4           NaN           NaN  23/1/2016                  10   

   COMPAGNIE AERIENNE  
0               NVPPA  
1                 COA  
2                  NA  
3                 THA  
4                 MAF  

[5 rows x 30 columns]

In [4]:
# Set the option to display all columns
pd.set_option('display.max_columns', None)
pd.set_option('display.expand_frame_repr', False)  # Prevent line wrapping
vols.describe()

IDENTIFIANT           VOL  DEPART PROGRAMME  HEURE DE DEPART  RETART DE DEPART  TEMPS DE DEPLACEMENT A TERRE AU DECOLLAGE     DECOLLAGE  TEMPS PROGRAMME   TEMPS PASSE  TEMPS DE VOL      DISTANCE  ATTERRISSAGE  TEMPS DE DEPLACEMENT A TERRE A L'ATTERRISSAGE  ARRIVEE PROGRAMMEE  HEURE D'ARRIVEE  RETARD A L'ARRIVEE  DETOURNEMENT    ANNULATION  RETARD SYSTEM  RETARD SECURITE  RETARD COMPAGNIE   RETARD AVION   RETARD METEO  NIVEAU DE SECURITE
count  3.000000e+06  3.000000e+06      3.000000e+06     2.950774e+06      2.950774e+06                               2.948930e+06  2.948930e+06     2.999994e+06  2.940415e+06  2.940415e+06  3.000000e+06  2.947124e+06                                   2.947124e+06        3.000000e+06     2.947124e+06        2.940415e+06  3.000000e+06  3.000000e+06  560231.000000    560231.000000     560231.000000  560231.000000  560231.000000           3000000.0
mean   2.616279e+06  2.274759e+03      1.328864e+03     1.333119e+03      9.455701e+00                               1.725790e+01  1.355781e+03     1.458505e+02  1.416684e+02  1.166334e+02  8.460305e+02  1.471310e+03                                   7.789882e+00        1.493656e+03     1.476963e+03        4.914760e+00  2.666333e-03  1.719533e-02      15.351337         0.087366         19.704245      22.785506       3.165969                10.0
std    1.727447e+06  1.824416e+03      4.854302e+02     4.980278e+02      3.914365e+01                               9.272093e+00  5.002325e+02     7.907005e+01  7.815676e+01  7.636377e+01  6.436554e+02  5.199641e+02                                   5.981326e+00        5.052021e+02     5.240664e+02        4.140800e+01  5.156767e-02  1.299987e-01      30.331968         2.327869         51.011396      44.240728      21.763745                 0.0
min    0.000000e+00  1.000000e+00      1.000000e+00     1.000000e+00     -8.200000e+01                               1.000000e+00  1.000000e+00     1.800000e+01  1.500000e+01  7.000000e+00  3.100000e+01  1.000000e+00                                   1.000000e+00        1.000000e+00     1.000000e+00       -8.700000e+01  0.000000e+00  0.000000e+00       0.000000         0.000000          0.000000       0.000000       0.000000                10.0
25%    1.043038e+06  7.800000e+02      9.160000e+02     9.200000e+02     -5.000000e+00                               1.200000e+01  9.360000e+02     8.800000e+01  8.500000e+01  6.000000e+01  3.700000e+02  1.055000e+03                                   4.000000e+00        1.110000e+03     1.100000e+03       -1.400000e+01  0.000000e+00  0.000000e+00       0.000000         0.000000          0.000000       0.000000       0.000000                10.0
50%    2.089116e+06  1.730000e+03      1.325000e+03     1.329000e+03     -2.000000e+00                               1.500000e+01  1.342000e+03     1.250000e+02  1.220000e+02  9.600000e+01  6.650000e+02  1.508000e+03                                   6.000000e+00        1.520000e+03     1.512000e+03       -5.000000e+00  0.000000e+00  0.000000e+00       4.000000         0.000000          1.000000       0.000000       0.000000                10.0
75%    4.288817e+06  3.414000e+03      1.730000e+03     1.738000e+03      7.000000e+00                               2.000000e+01  1.752000e+03     1.790000e+02  1.750000e+02  1.480000e+02  1.089000e+03  1.910000e+03                                   9.000000e+00        1.917000e+03     1.916000e+03        8.000000e+00  0.000000e+00  0.000000e+00      20.000000         0.000000         19.000000      28.000000       0.000000                10.0
max    5.332913e+06  9.855000e+03      2.359000e+03     2.400000e+03      1.988000e+03                               2.000000e+02  2.400000e+03     7.180000e+02  7.660000e+02  6.900000e+02  4.983000e+03  2.400000e+03                                   2.480000e+02        2.359000e+03     2.400000e+03        1.971000e+03  1.000000e+00  1.000000e+00     991.000000       440.000000       1971.000000    1313.000000   

En examinant les résultats de la sortie, on peut voir que la plupart des variables ont des valeurs minimales, maximales, moyennes et écarts-types raisonnables. Cependant, il y a des valeurs qui semblent être aberrantes et qui peuvent être des erreurs de saisie, comme le temps de déplacement à terre au décollage qui peut être négatif.

On peut également constater que la variable "NIVEAU DE SECURITE" a une valeur constante de 10, ce qui indique qu'elle n'apporte pas d'informations utiles pour notre analyse et peut être supprimée.

Enfin, on peut noter que certaines variables telles que "RETARD SYSTEM", "RETARD SECURITE", "RETARD COMPAGNIE", "RETARD AVION" et "RETARD METEO" ont une grande proportion de valeurs manquantes (environ 80% des données), mais il n'y a pas de valeurs manquantes dans les données, mais plutôt qu'il n'y a pas eu de retards de vol

In [7]:
import pandas as pd
import sqlite3

# Connexion à la base de données
conn = sqlite3.connect('/Users/Radja/Downloads/batch_1.db')

# Récupération des données de la table aeroports
compagnies = pd.read_sql_query("SELECT * from compagnies", conn)

# Fermeture de la connexion à la base de données
conn.close()

# Affichage des données
compagnies.head()

COMPAGNIE  CODE
0           Try Hard Airlines   THA
1  Corporate Overlord Airways   COA
2          Neverland Airlines    NA
3    Always A Problem Flights  AAPF
4              Overpriced Air    OA

In [8]:
compagnies.describe()

COMPAGNIE CODE
count                  13   13
unique                 13   13
top     Try Hard Airlines  THA
freq                    1    1

Le résultat affiché montre deux colonnes: "COMPAGNIE" et "CODE", avec des statistiques pour chaque colonne. Pour la colonne "COMPAGNIE", il y a 13 valeurs différentes (unique) sur un total de 13 (count). 

In [9]:
vols.rename(columns={'AEROPORT DEPART': 'AEROPORT DEPART A', 'AEROPORT ARRIVEE': 'AEROPORT ARRIVEE A' }, inplace=True)


In [10]:
# Jointure LEFT JOIN
merged_df = vols.merge(aéroports, left_on='AEROPORT DEPART A', right_on='CODE IATA', how='left')

In [11]:
row_count = len(merged_df)
print("Number of rows:", row_count)

Number of rows: 3000000


In [12]:
row_count2 = len(vols)
print("Number of rows:", row_count2)

Number of rows: 3000000


In [13]:
row_count3 = len(aéroports)
print("Number of rows:", row_count3)

Number of rows: 319


In [14]:
# Suppression des colonnes
columns_to_drop = ['CODE AVION','LONGITUDE', 'LATITUDE','HAUTEUR']
df = merged_df.drop(columns=columns_to_drop)

In [15]:
df.rename(columns={'NOM': 'AEROPORT DEPART', 'LIEU': 'LIEU DE DEPART', 'PAYS': 'PAYS DE DEPART', 'PRIX RETARD PREMIERE 10 MINUTES': 'PRIX RETARD PREMIERE 10 MINUTES DEPART','PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES': 'PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES DEPART' }, inplace=True)


In [16]:
# Jointure LEFT JOIN
merged_df = df.merge(aéroports, left_on='AEROPORT ARRIVEE A', right_on='CODE IATA', how='left')

In [17]:
# Suppression des colonnes
columns_to_drop = ["LONGITUDE", "LATITUDE","HAUTEUR","TEMPS DE DEPLACEMENT A TERRE AU DECOLLAGE", "TEMPS DE DEPLACEMENT A TERRE A L'ATTERRISSAGE"]
df = merged_df.drop(columns=columns_to_drop)

In [18]:
df.rename(columns={'CODE IATA_x':'CODE IATA DEPART','CODE IATA_y':'CODE IATA ARRIVEE', 'NOM': 'AEROPORT ARRIVEE', 'LIEU': 'LIEU DE ARRIVEE', 'PAYS': 'PAYS DARRIVEE', 'PRIX RETARD PREMIERE 10 MINUTES': 'PRIX RETARD PREMIERE 10 MINUTES ARRIVEE','PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES': 'PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES ARRIVEE' }, inplace=True)


In [19]:
# Suppression des colonnes
columns_to_drop = ['AEROPORT DEPART A', 'AEROPORT ARRIVEE A']
df = df.drop(columns=columns_to_drop)



In [20]:
# Set the option to display all columns
pd.set_option('display.max_columns', None)
pd.set_option('display.expand_frame_repr', False)  # Prevent line wrapping
df

IDENTIFIANT   VOL  DEPART PROGRAMME  HEURE DE DEPART  RETART DE DEPART  DECOLLAGE  TEMPS PROGRAMME  TEMPS PASSE  TEMPS DE VOL  DISTANCE  ATTERRISSAGE  ARRIVEE PROGRAMMEE  HEURE D'ARRIVEE  RETARD A L'ARRIVEE  DETOURNEMENT  ANNULATION RAISON D'ANNULATION  RETARD SYSTEM  RETARD SECURITE  RETARD COMPAGNIE  RETARD AVION  RETARD METEO       DATE  NIVEAU DE SECURITE COMPAGNIE AERIENNE CODE IATA DEPART                             AEROPORT DEPART LIEU DE DEPART PAYS DE DEPART  PRIX RETARD PREMIERE 10 MINUTES DEPART  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES DEPART CODE IATA ARRIVEE                     AEROPORT ARRIVEE    LIEU DE ARRIVEE PAYS DARRIVEE  PRIX RETARD PREMIERE 10 MINUTES ARRIVEE  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES ARRIVEE
0             651373  2317              1410           1411.0               1.0     1431.0            229.0        227.0         197.0      1749        2048.0                2059           2058.0                -1.0             0           0                None            NaN              NaN               NaN           NaN           NaN  10/4/2016                  10              NVPPA              TIA  Tirana International Airport Mother Teresa         Tirana             AL                                    4582                                                586                    DSS  Blaise Diagne International Airport              Dakar            SN                                     5499                                                120      
1            4153042  1755              1455           1449.0              -6.0     1504.0            210.0        191.0         173.0      1325        1857.0                1925           1900.0               -25.0             0           0                None            NaN              NaN               NaN           NaN           NaN  16/7/2018                  10                COA              MPM                              Maputo Airport         Maputo             MZ                                    4778                                                 12                    BOH                  Bournemouth Airport        Bournemouth            GB                                     3147                                                278      
2            3617513   804              1540           1607.0              27.0     1616.0            235.0        220.0         205.0      1520        1841.0                1835           1847.0                12.0             0           0                None            NaN              NaN               NaN           NaN           NaN   6/1/2018                  10                 NA              TRN                               Turin Airport         Torino             IT                                     352                                                849                    TNA                     Yaoqiang Airport              Jinan            CN                                     2066                                                795      
3            1829601   263              2226           2219.0              -7.0     2233.0             84.0         69.0          50.0       337        2323.0                2350           2328.0               -22.0             0           0                None            NaN              NaN               NaN           NaN           NaN   9/1/2016                  10                THA              BRU                            Brussels Airport       Brussels             BE                                    5077                                                 48                    SXF           Berlin-Schönefeld Airport             Berlin            DE                                     2151                                                685      
4            1250864  4511              1730           1719.0             -11.0     1733.0             71.0         66.0          45.0       252        1818.0                1841           1825.

In [21]:
# Calcul de la moyenne des retards
df['RETARD MOYEN'] = (df["RETART DE DEPART"] + df["RETARD A L'ARRIVEE"]) / 2


In [22]:
# Suppression des colonnes
columns_to_drop = [ "RETART DE DEPART", "RETARD A L'ARRIVEE" ]
df = df.drop(columns=columns_to_drop)

In [23]:
import pandas as pd

# Convertir la colonne 'DATE' en type de données de date/heure
df['DATE'] = pd.to_datetime(df['DATE'])

# Vérifier le nouveau type de données de la colonne 'DATE'
print(df['DATE'].dtype)


/var/folders/3t/tnl3lyzn58dczffhgsm53_gm0000gp/T/ipykernel_762/1196816865.py:4: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  df['DATE'] = pd.to_datetime(df['DATE'])


datetime64[ns]


In [24]:
print(df.dtypes)


IDENTIFIANT                                                         int64
VOL                                                                 int64
DEPART PROGRAMME                                                    int64
HEURE DE DEPART                                                   float64
DECOLLAGE                                                         float64
TEMPS PROGRAMME                                                   float64
TEMPS PASSE                                                       float64
TEMPS DE VOL                                                      float64
DISTANCE                                                            int64
ATTERRISSAGE                                                      float64
ARRIVEE PROGRAMMEE                                                  int64
HEURE D'ARRIVEE                                                   float64
DETOURNEMENT                                                        int64
ANNULATION                            

In [25]:

import pandas as pd

# Sélectionner les colonnes catégorielles à extraire
cat_cols = ['RAISON D\'ANNULATION', 'COMPAGNIE AERIENNE', 'CODE IATA DEPART','CODE IATA ARRIVEE']

# Extraire les colonnes catégorielles dans un nouveau DataFrame
cat_df = df.loc[:, cat_cols]

# Afficher le nouveau DataFrame contenant les colonnes catégorielles
cat_df



RAISON D'ANNULATION COMPAGNIE AERIENNE CODE IATA DEPART CODE IATA ARRIVEE
0                      None              NVPPA              TIA               DSS
1                      None                COA              MPM               BOH
2                      None                 NA              TRN               TNA
3                      None                THA              BRU               SXF
4                      None                MAF              VCE               AAL
...                     ...                ...              ...               ...
2999995                None              NVPPA              AGP               GOI
2999996                None                THA              DUR               MPM
2999997                None                THA              BOH               HAN
2999998                None              SMITH              DXB               PMI
2999999                None              NVPPA              VIE               BRU

[3000000 rows x 4 columns]

In [26]:
print(cat_df.dtypes)

RAISON D'ANNULATION    object
COMPAGNIE AERIENNE     object
CODE IATA DEPART       object
CODE IATA ARRIVEE      object
dtype: object


In [27]:
from sklearn.preprocessing import LabelEncoder

# Create an instance of LabelEncoder
label_encoder = LabelEncoder()

# Iterate over each categorical variable
for column in ['RAISON D\'ANNULATION', 'COMPAGNIE AERIENNE', 'CODE IATA DEPART', 'CODE IATA ARRIVEE']:
    # Fit the LabelEncoder on the column
    label_encoder.fit(cat_df[column])
    
    # Transform the column using LabelEncoder
    cat_df[column] = label_encoder.transform(cat_df[column])



In [28]:
print(cat_df)

         RAISON D'ANNULATION  COMPAGNIE AERIENNE  CODE IATA DEPART  CODE IATA ARRIVEE
0                          4                   9               280                 86
1                          4                   4               201                 32
2                          4                   8               288                284
3                          4                  12                38                269
4                          4                   7               302                  0
...                      ...                 ...               ...                ...
2999995                    4                   9                 7                115
2999996                    4                  12                89                201
2999997                    4                  12                32                124
2999998                    4                  11                92                232
2999999                    4                   9      

In [29]:
# Suppression des colonnes
columns_to_drop = ['RAISON D\'ANNULATION', 'COMPAGNIE AERIENNE', 'CODE IATA DEPART', 'CODE IATA ARRIVEE', 'AEROPORT DEPART', 'AEROPORT ARRIVEE',
          'LIEU DE DEPART', 'PAYS DE DEPART', 'LIEU DE ARRIVEE', 'PAYS DARRIVEE'] 
df_model = df.drop(columns=columns_to_drop)
df_model

IDENTIFIANT   VOL  DEPART PROGRAMME  HEURE DE DEPART  DECOLLAGE  TEMPS PROGRAMME  TEMPS PASSE  TEMPS DE VOL  DISTANCE  ATTERRISSAGE  ARRIVEE PROGRAMMEE  HEURE D'ARRIVEE  DETOURNEMENT  ANNULATION  RETARD SYSTEM  RETARD SECURITE  RETARD COMPAGNIE  RETARD AVION  RETARD METEO       DATE  NIVEAU DE SECURITE  PRIX RETARD PREMIERE 10 MINUTES DEPART  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES DEPART  PRIX RETARD PREMIERE 10 MINUTES ARRIVEE  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES ARRIVEE  RETARD MOYEN
0             651373  2317              1410           1411.0     1431.0            229.0        227.0         197.0      1749        2048.0                2059           2058.0             0           0            NaN              NaN               NaN           NaN           NaN 2016-10-04                  10                                    4582                                                586                                          5499                                                120                 0.0
1            4153042  1755              1455           1449.0     1504.0            210.0        191.0         173.0      1325        1857.0                1925           1900.0             0           0            NaN              NaN               NaN           NaN           NaN 2018-07-16                  10                                    4778                                                 12                                          3147                                                278               -15.5
2            3617513   804              1540           1607.0     1616.0            235.0        220.0         205.0      1520        1841.0                1835           1847.0             0           0            NaN              NaN               NaN           NaN           NaN 2018-06-01                  10                                     352                                                849                                          2066                                                795                19.5
3            1829601   263              2226           2219.0     2233.0             84.0         69.0          50.0       337        2323.0                2350           2328.0             0           0            NaN              NaN               NaN           NaN           NaN 2016-09-01                  10                                    5077                                                 48                                          2151                                                685               -14.5
4            1250864  4511              1730           1719.0     1733.0             71.0         66.0          45.0       252        1818.0                1841           1825.0             0           0            NaN              NaN               NaN           NaN           NaN 2016-01-23                  10                                    2143                                                434                                          8346                                                353               -13.5
...              ...   ...               ...              ...        ...              ...          ...           ...       ...           ...                 ...              ...           ...         ...            ...              ...               ...           ...           ...        ...                 ...                                     ...                                                ...                                           ...                                                ...                 ...
2999995       782415   769              2216           2216.0     2232.0            130.0        120.0         100.0       669        2312.0                2326           2316.0             0           0            NaN              NaN               NaN           NaN           NaN 2018-03-30                  10                                     771          

In [30]:
import pandas as pd

# Concatenate df_model and cat_df
df_concatenated = pd.concat([df_model, cat_df], axis=1)

df_concatenated


IDENTIFIANT   VOL  DEPART PROGRAMME  HEURE DE DEPART  DECOLLAGE  TEMPS PROGRAMME  TEMPS PASSE  TEMPS DE VOL  DISTANCE  ATTERRISSAGE  ARRIVEE PROGRAMMEE  HEURE D'ARRIVEE  DETOURNEMENT  ANNULATION  RETARD SYSTEM  RETARD SECURITE  RETARD COMPAGNIE  RETARD AVION  RETARD METEO       DATE  NIVEAU DE SECURITE  PRIX RETARD PREMIERE 10 MINUTES DEPART  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES DEPART  PRIX RETARD PREMIERE 10 MINUTES ARRIVEE  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES ARRIVEE  RETARD MOYEN  RAISON D'ANNULATION  COMPAGNIE AERIENNE  CODE IATA DEPART  CODE IATA ARRIVEE
0             651373  2317              1410           1411.0     1431.0            229.0        227.0         197.0      1749        2048.0                2059           2058.0             0           0            NaN              NaN               NaN           NaN           NaN 2016-10-04                  10                                    4582                                                586                                          5499                                                120                 0.0                    4                   9               280                 86
1            4153042  1755              1455           1449.0     1504.0            210.0        191.0         173.0      1325        1857.0                1925           1900.0             0           0            NaN              NaN               NaN           NaN           NaN 2018-07-16                  10                                    4778                                                 12                                          3147                                                278               -15.5                    4                   4               201                 32
2            3617513   804              1540           1607.0     1616.0            235.0        220.0         205.0      1520        1841.0                1835           1847.0             0           0            NaN              NaN               NaN           NaN           NaN 2018-06-01                  10                                     352                                                849                                          2066                                                795                19.5                    4                   8               288                284
3            1829601   263              2226           2219.0     2233.0             84.0         69.0          50.0       337        2323.0                2350           2328.0             0           0            NaN              NaN               NaN           NaN           NaN 2016-09-01                  10                                    5077                                                 48                                          2151                                                685               -14.5                    4                  12                38                269
4            1250864  4511              1730           1719.0     1733.0             71.0         66.0          45.0       252        1818.0                1841           1825.0             0           0            NaN              NaN               NaN           NaN           NaN 2016-01-23                  10                                    2143                                                434                                          8346                                                353               -13.5                    4                   7               302                  0
...              ...   ...               ...              ...        ...              ...          ...           ...       ...           ...                 ...              ...           ...         ...            ...              ...               ...           ...           ...        ...                 ...                                     ...                                                ...            

In [31]:
df_concatenated['RETARD SYSTEM'] = df_concatenated['RETARD SYSTEM'].fillna(0)
df_concatenated['RETARD SECURITE'] = df_concatenated['RETARD SECURITE'].fillna(0)
df_concatenated['RETARD COMPAGNIE'] = df_concatenated['RETARD COMPAGNIE'].fillna(0)
df_concatenated['RETARD AVION'] = df_concatenated['RETARD AVION'].fillna(0)
df_concatenated['RETARD METEO'] = df_concatenated['RETARD METEO'].fillna(0)
df_concatenated

IDENTIFIANT   VOL  DEPART PROGRAMME  HEURE DE DEPART  DECOLLAGE  TEMPS PROGRAMME  TEMPS PASSE  TEMPS DE VOL  DISTANCE  ATTERRISSAGE  ARRIVEE PROGRAMMEE  HEURE D'ARRIVEE  DETOURNEMENT  ANNULATION  RETARD SYSTEM  RETARD SECURITE  RETARD COMPAGNIE  RETARD AVION  RETARD METEO       DATE  NIVEAU DE SECURITE  PRIX RETARD PREMIERE 10 MINUTES DEPART  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES DEPART  PRIX RETARD PREMIERE 10 MINUTES ARRIVEE  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES ARRIVEE  RETARD MOYEN  RAISON D'ANNULATION  COMPAGNIE AERIENNE  CODE IATA DEPART  CODE IATA ARRIVEE
0             651373  2317              1410           1411.0     1431.0            229.0        227.0         197.0      1749        2048.0                2059           2058.0             0           0            0.0              0.0               0.0           0.0           0.0 2016-10-04                  10                                    4582                                                586                                          5499                                                120                 0.0                    4                   9               280                 86
1            4153042  1755              1455           1449.0     1504.0            210.0        191.0         173.0      1325        1857.0                1925           1900.0             0           0            0.0              0.0               0.0           0.0           0.0 2018-07-16                  10                                    4778                                                 12                                          3147                                                278               -15.5                    4                   4               201                 32
2            3617513   804              1540           1607.0     1616.0            235.0        220.0         205.0      1520        1841.0                1835           1847.0             0           0            0.0              0.0               0.0           0.0           0.0 2018-06-01                  10                                     352                                                849                                          2066                                                795                19.5                    4                   8               288                284
3            1829601   263              2226           2219.0     2233.0             84.0         69.0          50.0       337        2323.0                2350           2328.0             0           0            0.0              0.0               0.0           0.0           0.0 2016-09-01                  10                                    5077                                                 48                                          2151                                                685               -14.5                    4                  12                38                269
4            1250864  4511              1730           1719.0     1733.0             71.0         66.0          45.0       252        1818.0                1841           1825.0             0           0            0.0              0.0               0.0           0.0           0.0 2016-01-23                  10                                    2143                                                434                                          8346                                                353               -13.5                    4                   7               302                  0
...              ...   ...               ...              ...        ...              ...          ...           ...       ...           ...                 ...              ...           ...         ...            ...              ...               ...           ...           ...        ...                 ...                                     ...                                                ...            

In [32]:
import pandas as pd

# Créez une copie du DataFrame pour éviter les modifications sur l'original
df_modified = df_concatenated.copy()

# Créez une nouvelle colonne "Retard" initialisée à 0
df_modified['Retard'] = 0

# Attribuez la valeur 1 aux lignes où la colonne a une valeur strictement inférieure à 0
df_modified.loc[df_modified['RETARD MOYEN'] < 0, 'Retard'] = 1

# Affichez le DataFrame modifié avec la nouvelle colonne "Retard"
df_modified


IDENTIFIANT   VOL  DEPART PROGRAMME  HEURE DE DEPART  DECOLLAGE  TEMPS PROGRAMME  TEMPS PASSE  TEMPS DE VOL  DISTANCE  ATTERRISSAGE  ARRIVEE PROGRAMMEE  HEURE D'ARRIVEE  DETOURNEMENT  ANNULATION  RETARD SYSTEM  RETARD SECURITE  RETARD COMPAGNIE  RETARD AVION  RETARD METEO       DATE  NIVEAU DE SECURITE  PRIX RETARD PREMIERE 10 MINUTES DEPART  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES DEPART  PRIX RETARD PREMIERE 10 MINUTES ARRIVEE  PRIS RETARD POUR CHAQUE MINUTE APRES 10 MINUTES ARRIVEE  RETARD MOYEN  RAISON D'ANNULATION  COMPAGNIE AERIENNE  CODE IATA DEPART  CODE IATA ARRIVEE  Retard
0             651373  2317              1410           1411.0     1431.0            229.0        227.0         197.0      1749        2048.0                2059           2058.0             0           0            0.0              0.0               0.0           0.0           0.0 2016-10-04                  10                                    4582                                                586                                          5499                                                120                 0.0                    4                   9               280                 86       0
1            4153042  1755              1455           1449.0     1504.0            210.0        191.0         173.0      1325        1857.0                1925           1900.0             0           0            0.0              0.0               0.0           0.0           0.0 2018-07-16                  10                                    4778                                                 12                                          3147                                                278               -15.5                    4                   4               201                 32       1
2            3617513   804              1540           1607.0     1616.0            235.0        220.0         205.0      1520        1841.0                1835           1847.0             0           0            0.0              0.0               0.0           0.0           0.0 2018-06-01                  10                                     352                                                849                                          2066                                                795                19.5                    4                   8               288                284       0
3            1829601   263              2226           2219.0     2233.0             84.0         69.0          50.0       337        2323.0                2350           2328.0             0           0            0.0              0.0               0.0           0.0           0.0 2016-09-01                  10                                    5077                                                 48                                          2151                                                685               -14.5                    4                  12                38                269       1
4            1250864  4511              1730           1719.0     1733.0             71.0         66.0          45.0       252        1818.0                1841           1825.0             0           0            0.0              0.0               0.0           0.0           0.0 2016-01-23                  10                                    2143                                                434                                          8346                                                353               -13.5                    4                   7               302                  0       1
...              ...   ...               ...              ...        ...              ...          ...           ...       ...           ...                 ...              ...           ...         ...            ...              ...               ...           ...           ...        ...                 ...                                     ...               

In [33]:
import pandas as pd

# Check for NaN values in df_modified
nan_check = df_modified.isnull().any()

# Print the columns with NaN values
columns_with_nan = nan_check[nan_check].index
print("Columns with NaN values:")
print(columns_with_nan)


Columns with NaN values:
Index(['HEURE DE DEPART', 'DECOLLAGE', 'TEMPS PROGRAMME', 'TEMPS PASSE',
       'TEMPS DE VOL', 'ATTERRISSAGE', 'HEURE D'ARRIVEE', 'RETARD MOYEN'],
      dtype='object')


In [34]:
import pandas as pd

# Calculate the percentage of NaN values in each column
nan_percentage = df_modified.isnull().mean() * 100

# Print the columns and their respective percentages of NaN values
print("Percentage of NaN values in each column:")
print(nan_percentage)


Percentage of NaN values in each column:
IDENTIFIANT                                                0.000000
VOL                                                        0.000000
DEPART PROGRAMME                                           0.000000
HEURE DE DEPART                                            1.640867
DECOLLAGE                                                  1.702333
TEMPS PROGRAMME                                            0.000200
TEMPS PASSE                                                1.986167
TEMPS DE VOL                                               1.986167
DISTANCE                                                   0.000000
ATTERRISSAGE                                               1.762533
ARRIVEE PROGRAMMEE                                         0.000000
HEURE D'ARRIVEE                                            1.762533
DETOURNEMENT                                               0.000000
ANNULATION                                                 0.000000
RETARD 

In [35]:
import pandas as pd

# Drop NaN values from the specified columns
df_modified.dropna(subset=['HEURE DE DEPART', 'DECOLLAGE', 'TEMPS PROGRAMME', 'TEMPS PASSE', 'TEMPS DE VOL', 'ATTERRISSAGE', "HEURE D'ARRIVEE", 'RETARD MOYEN'], inplace=True)

# Verify if the NaN values have been dropped
nan_check = df_modified.isnull().any()
columns_with_nan = nan_check[nan_check].index
print("Columns with NaN values after dropping:")
print(columns_with_nan)


Columns with NaN values after dropping:
Index([], dtype='object')


In [49]:
from sklearn.model_selection import train_test_split

# Séparez les variables d'entrée (X) et la variable cible (y)
X = df_modified.drop(columns=['Retard'])
y = df_modified['Retard']

# Divisez les données en ensembles d'entraînement et de test
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Affichez les dimensions des ensembles de données
print("Dimensions de l'ensemble d'entraînement : ", x_train.shape, y_train.shape)
print("Dimensions de l'ensemble de test : ", x_test.shape, y_test.shape)


Dimensions de l'ensemble d'entraînement :  (2352332, 30) (2352332,)
Dimensions de l'ensemble de test :  (588083, 30) (588083,)


In [50]:
# Liste des colonnes à exclure
columns_to_exclude = ['DATE']

# Exclusion des colonnes de l'ensemble d'entraînement et de test
x_train = x_train.drop(columns=columns_to_exclude)
x_test = x_test.drop(columns=columns_to_exclude)


In [104]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Create the logistic regression model
model = LogisticRegression(max_iter=1000)

# Train the model on the training data
model.fit(x_train, y_train)

# Make predictions on the test data
y_pred = model.predict(x_test)

# Compute evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Print the evaluation metrics
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)


/Users/Radja/Library/Python/3.8/lib/python/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Accuracy: 0.8066803495424966
Precision: 0.7627339008766015
Recall: 0.9969949373922337
F1-score: 0.8642716095958282


Les résultats du modèle de régression logistique sur les données de test sont les suivants :

Accuracy : 0,8067
L'accuracy représente la proportion d'étiquettes correctement prédites par rapport au nombre total d'instances dans l'ensemble de test. Dans ce cas, le modèle a atteint une précision de 0,8067, ce qui signifie qu'il a correctement prédit la classe d'environ 80,67 % des instances dans l'ensemble de test.


Précision : 0,7627
La précision mesure la proportion d'instances positives correctement prédites (vrais positifs) par rapport au nombre total d'instances prédites comme positives (vrais positifs + faux positifs). Une précision de 0,7627 indique que parmi toutes les instances prédites comme positives, environ 76,27 % d'entre elles étaient réellement positives.


Rappel : 0,9970
Le rappel, également connu sous le nom de sensibilité ou taux de vrais positifs, mesure la proportion d'instances positives correctement prédites (vrais positifs) par rapport au nombre total d'instances positives réelles (vrais positifs + faux négatifs). Un rappel de 0,9970 suggère que le modèle a correctement identifié environ 99,70 % des instances positives réelles.


Score F1 : 0,8643
Le score F1 est la moyenne harmonique de la précision et du rappel. Il fournit une mesure équilibrée de la performance du modèle en tenant compte à la fois de la précision et du rappel. Un score F1 de 0,8643 indique un bon équilibre entre la précision et le rappel, les valeurs plus élevées indiquant une meilleure performance globale.



Dans l'ensemble, le modèle présente une performance relativement bonne avec un rappel élevé (capacité à identifier les instances positives) mais une précision légèrement plus basse (exactitude des prédictions positives). Cela signifie que le modèle peut générer quelques faux positifs, mais il est efficace pour capturer la plupart des instances positives. Le score F1 confirme la bonne performance globale du modèle en tenant compte à la fois de la précision et du rappel.

In [37]:
from sklearn.model_selection import train_test_split

# Sélectionnez un échantillon de 10% de df_modified
sample_df = df_modified.sample(frac=0.01, random_state=42)

# Séparez les variables d'entrée (X) et la variable cible (y)
X = sample_df.drop(columns=['Retard'])
y = sample_df['Retard']

# Divisez les données en ensembles d'entraînement et de test
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Affichez les dimensions des ensembles de données
print("Dimensions de l'ensemble d'entraînement : ", x_train.shape, y_train.shape)
print("Dimensions de l'ensemble de test : ", x_test.shape, y_test.shape)

# Liste des colonnes à exclure
columns_to_exclude = ['DATE']

# Exclusion des colonnes de l'ensemble d'entraînement et de test
x_train = x_train.drop(columns=columns_to_exclude)
x_test = x_test.drop(columns=columns_to_exclude)


Dimensions de l'ensemble d'entraînement :  (23523, 30) (23523,)
Dimensions de l'ensemble de test :  (5881, 30) (5881,)


In [41]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Create an SVM classifier object with adjusted parameters
model = SVC(kernel='rbf', C=1.0)

# Train the model on the training data
model.fit(x_train, y_train)

# Make predictions on the test data
y_pred = model.predict(x_test)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Print the evaluation metrics
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)


Accuracy: 0.6087400102023466
Precision: 0.6087400102023466
Recall: 1.0
F1-score: 0.7567910368882782


These metrics indicate the performance of your model on the test data. The accuracy of 0.6087 means that approximately 60.87% of the predictions made by the model are correct. Precision of 0.6087 indicates that out of all the positive predictions made by the model, 60.87% are true positives. Recall of 1.0 suggests that the model is able to identify all the actual positive instances correctly. The F1-score of 0.7568 is the harmonic mean of precision and recall, providing an overall measure of the model's performance.



In [48]:
import numpy as np
from tensorflow import keras
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from tensorflow.keras.callbacks import EarlyStopping

# Convert your training and test data to numpy arrays (if not already)
x_train = np.array(x_train)
y_train = np.array(y_train)
x_test = np.array(x_test)
y_test = np.array(y_test)

# Define the architecture of the neural network
model = keras.Sequential([
    keras.layers.Dense(64, activation='relu', input_dim=x_train.shape[1]),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Define the early stopping criteria
early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)

# Train the model with early stopping
model.fit(x_train, y_train, epochs=100, batch_size=32, validation_data=(x_test, y_test), callbacks=[early_stopping])

# Make predictions on the test data
y_pred_prob = model.predict(x_test)
y_pred = np.where(y_pred_prob >= 0.5, 1, 0)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Print the evaluation metrics
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)



Epoch 1/100
736/736 [==============================] - 3s 2ms/step - loss: 5462.4707 - accuracy: 0.5220 - val_loss: 318.3347 - val_accuracy: 0.5810
Epoch 2/100
736/736 [==============================] - 1s 2ms/step - loss: 2268.2417 - accuracy: 0.5355 - val_loss: 57.5372 - val_accuracy: 0.6368
Epoch 3/100
736/736 [==============================] - 1s 2ms/step - loss: 2020.9735 - accuracy: 0.5414 - val_loss: 1129.6492 - val_accuracy: 0.4084
Epoch 4/100
736/736 [==============================] - 1s 2ms/step - loss: 2065.8494 - accuracy: 0.5415 - val_loss: 7209.3340 - val_accuracy: 0.3923
Epoch 5/100
736/736 [==============================] - 1s 2ms/step - loss: 1925.5896 - accuracy: 0.5483 - val_loss: 188.5802 - val_accuracy: 0.6589
Epoch 6/100
736/736 [==============================] - 1s 2ms/step - loss: 1345.2651 - accuracy: 0.5584 - val_loss: 53.9863 - val_accuracy: 0.7218
Epoch 7/100
736/736 [==============================] - 1s 2ms/step - loss: 1175.8671 - accuracy: 0.5664 - val_lo

Le processus d'entraînement s'est déroulé sur 16 époques. Chaque époque représente un passage complet à travers les données d'entraînement.
La perte d'entraînement (représentée par "loss") et l'exactitude d'entraînement ont été rapportées pour chaque époque. Ces valeurs indiquent à quel point le modèle s'est adapté aux données d'entraînement à chaque époque. On peut observer que la perte d'entraînement a diminué progressivement au fil des époques, tandis que l'exactitude d'entraînement a augmenté.
La perte de validation (représentée par "val_loss") et l'exactitude de validation ont également été rapportées pour chaque époque. Ces valeurs indiquent à quel point le modèle généralisait aux données invisibles (c'est-à-dire l'ensemble de validation). On peut constater que la perte de validation et l'exactitude ont fluctué au cours des époques.
Les métriques d'évaluation finales sur les données de test ont également été fournies :
Exactitude : 0,7218
Précision : 0,7258
Rappel : 0,8726
F1-score : 0,7925
Ces métriques fournissent une mesure de la performance du modèle sur les données de test. L'exactitude représente la justesse globale des prédictions, tandis que la précision, le rappel et le F1-score donnent des indications sur les performances du modèle pour les prédictions de la classe positive. Dans ce cas, le modèle a obtenu une performance modérée à bonne avec une exactitude de 0,7218 et une précision, un rappel et un F1-score équilibrés.